### ISSUE TITLE
Confusing behavior typechecking array members

### ISSUE THREAD

**Issue by cwmoo740** [microsoft/TypeScript#14828]
**TypeScript Version:**  2.2.1

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
interface Thing {
    prop: Record<string, string>;
}

const bug: Thing[] = [
    { prop: {} },
    { prop: 5 }
];

const also_bug: Thing[] = [
    { prop: ['5'] },
    { prop: {} },
    { prop: {inner: '5'} },
    { prop: []}
];

// adding an inner property makes it work
const caught: Thing[] = [
    { prop: '5' },
    { prop: { innerprop: 'string' } }
];
```

**Expected behavior:**
tsc should report errors for bug, also_bug, and caught.
**Actual behavior:**
tsc only reports an error for caught.

An array of objects, where the inner property is an empty object, does not seem to typecheck properly.

The same example can be found here: 

https://github.com/cwmoo740/typescript-bug

**Comment by RyanCavanaugh** [comment]
You can see what's going on by writing this:
```ts
// Inferred type: qq: { prop: {} }[]
const qq = [
    { prop: {} },
    { prop: 5 }
];
```
The type of an array is determined by the union of its elements, and a union type `T | U` will eagerly reduce to `T` if `U` is a subtype of `T`. e.g.:
```ts
// dogs: Dog[], not Array<Dog | Terrier | Poodle>
const dogs = [someDog, terrier, poodle];
```

Two things here:
 * When an array literal is contextually typed by an array type, we *shouldn't* allow subtype reduction of its element type
 * I'm not convinced `5` should be a subtype of `{ }` anymore. `{ }` should have the type `object` now and stop reduction

@ahejlsberg thoughts?

**Comment by ahejlsberg** [comment]
> When an array literal is contextually typed by an array type, we shouldn't allow subtype reduction of its element type

That seems reasonable. It would be a fairly small change in `checkArrayLiteral`.

> I'm not convinced 5 should be a subtype of { } anymore.

For backward compatibility reasons the primitive types have to be subtypes of the `{}` type.

> { } should have the type object now and stop reduction

I guess we could experiment with giving type `object` to the expression `{}`. But I have a hunch it will break stuff too.

**Comment by RyanCavanaugh** [comment]
> the primitive types have to be subtypes of the {} type.

To clarify, I meant that the type of the literal `5` should not be a subtype of the type of the expression `{ }` (by some means - either giving `{ }` the type `object`, or something else)

**Comment by RyanCavanaugh** [comment]
Tentative Bug tag to try the contextual typing change to see how it goes

**Comment by Andarist** [comment]
@RyanCavanaugh this is an ancient issue - it works in all TS versions that are available in the TS playground.

**Comment by RyanCavanaugh** [comment]
@Andarist 🙏
